### PR TITLE
Remove `Return::Unit` variant

### LINKED ISSUE #15447: Consider removing `Return::Unit`
## What problem does this solve or what need does it fill?

Currently the return type of `DynamicFunction` has four variants:
```
pub enum Return<'a> {
    Unit,
    Owned(Box<dyn PartialReflect>),
    Ref(&'a dyn PartialReflect),
    Mut(&'a mut dyn PartialReflect),
}
```

In my opinion a small improvement stop special casing the unit variant and instead just return the equivalent of `()`. This is analogous to rust functions returning `()` instead of some badly-defined `void` "type" like in C++.

## What solution would you like?

- Remove the `Unit` variant
- Instead use `Return::Owned(Box::new(()))`
- Update `Return::is_unit` to check if it's an owned value holding a reflected value representing a unit

## What alternative(s) have you considered?

In my code I currently just have a special case for the unit variant:
```
// where `ret: Return`
if ret.is_unit() {
    ().into()
} else {
    ret.unwrap_owned()
}
```

### PR DESCRIPTION
**Pull request by hooded-shrimp**
# Objective

- Fixes #15447 

## Solution

- Remove the `Return::Unit` variant and use a `Return::Owned` variant holding a unit `()` type.

## Migration Guide

- Removed the `Return::Unit` variant; use `Return::unit()` instead.

### CODE REVIEW

File: crates/bevy_reflect/src/func/return_type.rs [hunk 1]
```diff
@@ -6,8 +6,6 @@ use crate::PartialReflect;
 /// [`DynamicFunctionMut`]: crate::func::DynamicFunctionMut
 #[derive(Debug)]
 pub enum Return<'a> {
-    /// The function returns nothing (i.e. it returns `()`).
-    Unit,
     /// The function returns an owned value.
```
Comment (pablo-lua) [inline crates/bevy_reflect/src/func/return_type.rs:9]: I think we need to warn the user that this value can be "nothing" (Like Return::Unit says above in the docs)

**Review by BenjaminBrienen — APPROVED** [review]

File: crates/bevy_reflect/src/func/return_type.rs (hunk 1 above)
Comment (hooded-shrimp) [reply crates/bevy_reflect/src/func/return_type.rs:9]: I added something to let users know.

I am not the best at writing documentation, so if it should be written in a different way, let me know,

File: crates/bevy_reflect/src/func/return_type.rs [hunk 2]
```diff
@@ -6,9 +6,9 @@ use crate::PartialReflect;
...
+    ///
+    /// Owned values can be unit (`()`).
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/func/return_type.rs:11]: Maybe something like:
```suggestion
    /// This includes functions that return nothing (i.e. they return `()`).
```

File: crates/bevy_reflect/src/func/return_type.rs [hunk 3]
```diff
@@ -17,9 +17,17 @@ pub enum Return<'a> {
 }
 
 impl<'a> Return<'a> {
-    /// Returns `true` if the return value is [`Self::Unit`].
+    /// Creates an [`Owned`](Self::Owned) unit (`()`) type.
+    pub fn unit() -> Self {
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/func/return_type.rs:21]: Personally I’d prefer an associated constant but I just looked it up and `Box::new` isn't const 😭

**Review by MrGVSV — APPROVED** [review]
One documentation nit, but otherwise looks good to me!

File: crates/bevy_reflect/src/func/return_type.rs (hunk 1 above)
Comment (MrGVSV) [reply crates/bevy_reflect/src/func/return_type.rs:9]: @hooded-shrimp gave a suggestion :)

**Review by BenjaminBrienen — APPROVED** [review]

**Review by pablo-lua — APPROVED** [review]

**Pull request merged**